### PR TITLE
Static router for Versal architecture

### PR DESCRIPTION
**Pull request by WenhaoLin-AMD**
Make the static router in RWRoute support Versal architecture.

### CODE REVIEW

File: src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java [hunk 1]
```diff
@@ -404,12 +421,28 @@ public static void routeStaticNet(Net currNet,
                         break;
                     }
 
+                    TileTypeEnum tileTypeEnum = node.getTile().getTileTypeEnum();
+                    // On Versal, only allow IRI -> BLI_CLE_BOT_CORE* routethrus
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java:425]:
```suggestion
                    // On Versal, only allow IRI routethrus on BLI_CLE_BOT_CORE* tile types
```

**Comment by eddieh-xlnx** [comment]
This PR is for routing static nets when the sink pins are known (which is mostly the case since we start from an fully routed design). Inferring those static sink pins from a completely unrouted design is out of scope here and will be addressed in a future PR.

The static net router is more or less the same as for UltraScale/UltraScale+, but with a few Versal-only exceptions (e.g. intent codes, static source pins, etc.)

File: src/com/xilinx/rapidwright/design/DesignTools.java [hunk 2]
```diff
@@ -3202,24 +3214,93 @@ public static void createCeClkOfRoutethruFFToVCC(Design design) {
...
+                }
+
+                // Construct site pin from BEL name (e.g. [A-H][65]LUT) and pin name (A[1-6])
+                String sitePinName = belName.charAt(0) + "6";
+                SitePinInst pin = si.getSitePinInst(sitePinName);
+                if (pin == null) {
+                    staticNet.createPin(sitePinName, si);
+                } else if (!pin.getNet().equals(staticNet)) {
+                    // pin.getNet().removePin(pin);
+                    // net.addPin(pin);
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/design/DesignTools.java:3302]:
```suggestion
```

File: test/src/com/xilinx/rapidwright/rwroute/TestGlobalSignalRouting.java [hunk 3]
```diff
@@ -104,30 +118,104 @@ public void testRouteStaticNet() {
...
+                    vccNet.removePin(spi);
+                    spi.detachSiteInst();
+                }
+            }
+
+            Assertions.assertEquals(123, gndPins.size());
+            Assertions.assertEquals(230, vccPins.size());
+        }
+
+        // Even though we're starting from a fully-routed design, Versal designs may still need
```
Comment (eddieh-xlnx) [inline test/src/com/xilinx/rapidwright/rwroute/TestGlobalSignalRouting.java:200]:
```suggestion
        // Even though we may be starting from a fully-routed design, Versal designs still need
```

File: test/src/com/xilinx/rapidwright/rwroute/TestGlobalSignalRouting.java [hunk 4]
```diff
@@ -86,6 +91,15 @@ public void testRAMB36(String logicalPinName, String erroringSitePinName) throws
         }
     }
 
+    NodeStatus getNodeState(Design design, NetType netType, Node n) {
```
Comment (eddieh-xlnx) [inline test/src/com/xilinx/rapidwright/rwroute/TestGlobalSignalRouting.java:94]:
```suggestion
    // NOTE: This method does not avoid any existing routing (static or signal),
    //       only their pins
    NodeStatus getNodeState(Design design, NetType netType, Node n) {
```

File: src/com/xilinx/rapidwright/design/DesignTools.java [hunk 5]
```diff
@@ -3234,70 +3325,147 @@ public static void createA1A6ToStaticNets(Design design) {
...
+                        continue;
+                    }
+
+                    Pair<String, String> sitePinNames = pinMapping.get(bel.getName());
+                    final String[] belPinNames = new String[] {"CE", "SR"};
+                    for (String belPinName : belPinNames) {
+                        String sitePinName = (belPinName == belPinNames[0]) ? sitePinNames.getFirst() : sitePinNames.getSecond();
+                        SitePinInst spi = si.getSitePinInst(sitePinName);
+                        if (spi != null) {
+                            if (belPinName == belPinNames[0]) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/DesignTools.java:3358]: `belPinNames[0]` could be replaced with `CE` if we use the refactoring above.  This makes for more readable code.

File: src/com/xilinx/rapidwright/design/DesignTools.java [hunk 6]
```diff
@@ -3234,70 +3325,147 @@ public static void createA1A6ToStaticNets(Design design) {
...
+                            }
+                            // SR
+                            if (!spi.getNet().isGNDNet()) {
+                                continue;
+                            }
+                        }
+
+                        Net net = si.getNetFromSiteWire(sitePinName);
+                        if (net != null) {
+                            if (belPinName == belPinNames[0]) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/DesignTools.java:3370]: `belPinNames[0]` could be replaced with `CE` if we use the refactoring above.  This makes for more readable code.

File: src/com/xilinx/rapidwright/design/DesignTools.java [hunk 7]
```diff
@@ -3234,70 +3325,147 @@ public static void createA1A6ToStaticNets(Design design) {
...
-                    for (String pinName : Arrays.asList("RSTREGBU", "RSTREGBL")) {
-                        maybeCreateVccPinAndPossibleInversion(si, pinName, vcc, gndInvertibleToVcc);
+            }
+
+            // Remove all pins that are no longer on the GND net
+            Net gndNet = design.getGndNet();
+            gndNet.getPins().removeIf(spi -> spi.getNet() != gndNet);
+        } else if (series == Series.UltraScale || series == Series.UltraScalePlus) {
+            Net gndInvertibleToVcc = design.getGndNet();
+            final String[] pins = new String[] {"CE", "SR"};
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/DesignTools.java:3412]: See the note above about pulling this out as a static variable.

File: src/com/xilinx/rapidwright/design/DesignTools.java [hunk 8]
```diff
@@ -3234,70 +3325,147 @@ public static void createA1A6ToStaticNets(Design design) {
...
-                    if (siteWire.endsWith("L_O")) {
-                        pinName = "RSTREGBL";
-                    } else {
-                        pinName = "RSTREGBU";
+                    BEL bel = cell.getBEL();
+                    Pair<String, String> sitePinNames = pinMapping.get(bel.getBELType());
+                    for (String pin : pins) {
+                        BELPin belPin = cell.getBEL().getPin(pin);
+                        Net net = si.getNetFromSiteWire(belPin.getSiteWireName());
+                        if (net == null || (net == gndInvertibleToVcc && pin.equals("SR"))) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/DesignTools.java:3424]: Could use a static version of `SR` here.

File: src/com/xilinx/rapidwright/design/DesignTools.java [hunk 9]
```diff
@@ -3234,70 +3325,147 @@ public static void createA1A6ToStaticNets(Design design) {
...
+                            String sitePinName;
+                            if (pin.equals("CE")) { // CKEN
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/design/DesignTools.java:3426]: Could use a static version of CE here.

**Review by clavin-xlnx — APPROVED** [review]

**Pull request merged**